### ISSUE TITLE
req/s in output for shapes

### ISSUE THREAD

**Issue by yabouelenein** [locustio/locust#2810]
### Prerequisites

- [X] I am using [the latest version of Locust](https://github.com/locustio/locust/releases/)
- [X] I am reporting a bug, not asking a question

### Description

I am running the following shape code (expected_run_time is 300 and buffer_to_spin_down_init is 5) exact code from https://github.com/locustio/locust/blob/master/examples/custom_shape/stages.py:
```
 stages = [
        {"duration": expected_run_time/2, "users": number_of_users, "spawn_rate": spawn_rate, "user_classes": [InitializeUser]},
        {"duration": expected_run_time/2 + buffer_to_spin_down_init, "users": 0, "spawn_rate": number_of_users, "user_classes": [InitializeUser]}, 
        {"duration": expected_run_time + buffer_to_spin_down_init, "users": number_of_users, "spawn_rate": spawn_rate, "user_classes": [EvaluateUser]}
        ]
```

So the expected run time is 300 seconds with InitializeUser running for about 150s and EvaluateUser is about 145s.

This is the output I get at the end of the run:

```
Type     Name                                                                          # reqs      # fails |    Avg     Min     Max    Med |   req/s  failures/s
--------|----------------------------------------------------------------------------|-------|-------------|-------|-------|-------|-------|--------|-----------
POST     Evaluate-Journey                                                               33006     0(0.00%) |     12       3     754     12 |  106.21        0.00
POST     Initialize-Journey                                                             31738     0(0.00%) |     13       4     925     13 |  102.13        0.00
--------|----------------------------------------------------------------------------|-------|-------------|-------|-------|-------|-------|--------|-----------
         Aggregated                                                                     64744     0(0.00%) |     13       3     925     13 |  208.33        0.00
```

As you can see from the above, the # reqs for the Evaluate-Journey is 33006 and the run time is 145 seconds so the req/s should be   227.62 and similar behaviour can be seen with the Initialize-Journey.

The Charts also show a similar average:
![image](https://github.com/user-attachments/assets/74e41eca-e52b-4e9b-b926-bc924bccab01)

### Command line

-f /mnt/locust/main.py,/mnt/locust/default_custom_shape.py

### Locustfile contents

```python3
I am using a shape very similar to the one found in https://github.com/locustio/locust/blob/master/examples/custom_shape/stages.py
```

### Python version

3.10.5

### Locust version

2.29.1

### Operating system

Windows

**Comment by cyberw** [comment]
I'm not sure I follow completely, but RPS is calculated for the whole duration of the test, so this is kind of expected?

In fact, shapes have no control/impact on how stats are calculated, it is the same as running regular ramp up.

Probably the best way to handle this is by running two separate tests.

**Comment by cyberw** [comment]
Yes, exactly. Unfortunately there's no way to ask Locust to calculate the RPS only for the part where a certain user type was spawned, and it would be very hard to implement in a sensible way (the stats module doesnt even know which User they came from, and even if it did, it has no real knowledge of when users were spawned/despawned).

**Comment by yukha-dw** [comment]
> Yes, exactly. Unfortunately there's no way to ask Locust to calculate the RPS only for the part where a certain user type was spawned, and it would be very hard to implement in a sensible way (the stats module doesnt even know which User they came from, and even if it did, it has no real knowledge of when users were spawned/despawned).

Hello @cyberw, I have similar issue with how the stats are recorded, yet I am not sure if it is worth to create new Issue.

Is it possible to pre-warm the locust by spawning requested number of users first before recording happened? On short test, RPS is often downsized pretty badly due to delay of spawning thousands of users.

My remedy is to run the test with `undefined` duration, then reset the stats after all users have spawned. But generating the report of specific amount of duration is hard to do. I hope we can automate this behavior on python script instead.
